test/integration/tests/metagov.ts

```diff
@@ -0,0 +1,457 @@
+      await contracts.angle.connect(signer).transfer(locker.address, e18(10_000_000));
+    });
+
+    describe('without METAGOVERNANCE_TOKEN_STAKING role', function () {
+      it('lock() should revert', async () => {
+        await expectRevert(locker.lock(), 'UNAUTHORIZED');
+      });
+
+      it('exitLock() should revert', async () => {
+        await expectRevert(locker.exitLock(), 'UNAUTHORIZED');
+      });
+    });
+
+    describe('with METAGOVERNANCE_TOKEN_STAKING role', function () {
+      before(async function () {
+        // grant role
+        await forceEth(contracts.feiDAOTimelock.address);
+        const daoSigner = await getImpersonatedSigner(contracts.feiDAOTimelock.address);
+        await contracts.core
+          .connect(daoSigner)
+          .grantRole(ethers.utils.id('METAGOVERNANCE_TOKEN_STAKING'), deployAddress);
+      });
+
+      after(async function () {
+        // revoke role
+        await forceEth(contracts.feiDAOTimelock.address);
+        const daoSigner = await getImpersonatedSigner(contracts.feiDAOTimelock.address);
+        await contracts.core
+          .connect(daoSigner)
+          .revokeRole(ethers.utils.id('METAGOVERNANCE_TOKEN_STAKING'), deployAddress);
+      });
+
+      it('initial lock()', async function () {
+        expect(await contracts.angle.balanceOf(locker.address)).to.be.equal(e18(10_000_000));
+        expect(await contracts.veAngle.balanceOf(locker.address)).to.be.equal('0');
+        await locker.lock();
+        expect(await contracts.angle.balanceOf(locker.address)).to.be.equal('0');
+        expect(await contracts.veAngle.balanceOf(locker.address)).to.be.at.least(e18(9_900_000));
+      });
+      it('2 years later, increase lock time', async function () {
+        await time.increase(2 * 365 * 86400); // 2 years
+        expect(await contracts.angle.balanceOf(locker.address)).to.be.equal('0');
+        expect(await contracts.veAngle.balanceOf(locker.address)).to.be.at.least(e18(4_900_000));
+        expect(await contracts.veAngle.balanceOf(locker.address)).to.be.at.most(e18(5_100_000));
+        await locker.lock();
+        expect(await contracts.angle.balanceOf(locker.address)).to.be.equal('0');
+        expect(await contracts.veAngle.balanceOf(locker.address)).to.be.at.least(e18(9_900_000));
+      });
+      it('lock more tokens', async function () {
+        const holder = '0x4f91F01cE8ec07c9B1f6a82c18811848254917Ab';
+        await forceEth(holder);
+        const signer = await getImpersonatedSigner(holder);
+        await contracts.angle.connect(signer).transfer(locker.address, e18(1_000_000));
+        await locker.lock();
+        expect(await contracts.angle.balanceOf(locker.address)).to.be.equal('0');
+        expect(await contracts.veAngle.balanceOf(locker.address)).to.be.at.least(e18(10_900_000));
+      });
+      it('fast-forward 4 years', async function () {
+        // fast-forwarding 4 years directly causes a revert without reason,
```

Comment:
Hmm, do we know why it causes a revert?

Reply:
I think it has to do with the veToken implementation, some mathematical formulas does not work properly if the checkpoints are too spaced in time, or something like that. It reverts inside the checkpoint() call of the veToken contract, not on our side